Initial commit of MQTT entity management framework.

skotl opened the pull request:
Addable extension method `AddMqttExtensions()`
Rough entity creation through `EntityUpdater` (which is probably a poor name) that needs more strongly-typed.

**Note that I can't push this to a branch on the netdaemon core, which I think it would be better to be in - if you want to create a new branch ("mqtt_extensions" or similar?) then I can update the PR to point to that instead.**

That will also allow Eugene and I to collaborate around the branch without upsetting your work.

This has been tested manually within the context of the `Create entities from ND` discord chat.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
Non-breaking change


## Proposed change
New extension project called `NetDaemon.Extensions.MqttEntities` that adds a services extension `AddMqttExtensions()` that can be called from program.cs

Sample usage:

```csharp
    public MqttTester(ILogger<MqttTester> logger, IEntityUpdater entityUpdater)
    {
        _logger = logger;
        _entityUpdater = entityUpdater;

        _entityUpdater.CreateAsync("binary_sensor", "motion", "motion_test_1", "mqtt test")
            .GetAwaiter();
    }
```

Requires additional config in `appSettings.json`:

```json
  "Mqtt": {
    "Host": "homeassistant.local",
    "Port": "1883",
    "UserId": "mqtt",
    "Password": "secret"
  }

```


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]


<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[docs-repository]: https://github.com/net-daemon/docs

FrankBakkerNl commented on NetDaemon.Extensions.MqttEntities/DependencyInjectionSetup.cs:
```diff
@@ -0,0 +1,20 @@
+    public static IServiceCollection AddMqttExtensions(this IServiceCollection services)
+    {
+        services.AddSingleton<IMqttFactory, MqttFactory>();
+        services.AddScoped<IEntityUpdater, EntityUpdater>();
```
Scoped is only usefull if you have state in tje class that should be scoped (per app in case of ND) I think thats not the case here.

skotl replied:
My concern is the connection to mqtt, which times out after 10 seconds.
But actually... I remake the connection each time when a message is called, so that's probably not an issue.

OK - will update - thanks

helto4real commented:
@skotl I will add you to the dev team here. You can push to branch here your self and make it easier for you guys to collaborate

skotl commented on NetDaemon.Extensions.MqttEntities/DependencyInjectionSetup.cs:
```diff
@@ -0,0 +1,20 @@
+    /// <param name="services">Provided service collection</param>
+    public static IServiceCollection AddMqttExtensions(this IServiceCollection services)
+    {
+        services.AddSingleton<IMqttFactory, MqttFactory>();
```
It's in the MQTTnet nuget package

skotl commented:
Changed DI scope to Singleton for the local services, changed target branch to mqtt_extensions.
Just need to check that @FrankBakkerNl  is happy with the comment on MqttFactory

BeeHiveJava commented on NetDaemon.Extensions.MqttEntities/IEntityUpdater.cs:
```diff
@@ -0,0 +1,6 @@
+
+public interface IEntityUpdater
+{
+    Task CreateAsync(string deviceType, string deviceClass, string entityId, string name);
```
Maybe it's worth it to create a value object here to avoid passing around many parameters.

BeeHiveJava replied:
Also cancellation token support?

skotl replied:
My ambition was to get the extension framework in so that Eugene could take on the actual implentation (which is likely to be of the `<T> : where T is Entity` approach) so I wouldn't take this implementation of `IEntityUpdater` as representing how it will end up.
It's really just intended as an example of how to call the MQTT message sender.

BeeHiveJava replied:
Fair enough!

BeeHiveJava commented on NetDaemon.Extensions.MqttEntities/IMessageSender.cs:
```diff
@@ -0,0 +1,6 @@
+
+public interface IMessageSender
+{
+    Task SendMessageAsync(string topic, string payload);
```
Cancellation token support?

skotl replied:
Do you think it's needed when the connection, message-send and disposal is all wrapped in a single atomic call?

BeeHiveJava commented on NetDaemon.Extensions.MqttEntities/MessageSender.cs:
```diff
@@ -0,0 +1,64 @@
+        _logger = logger;
+        _mqttFactory = mqttFactory;
+        
+        _mqttConfig = configuration.GetSection("Mqtt")
```
Should use the options framework instead here.

BeeHiveJava commented on NetDaemon.Extensions.MqttEntities/MessageSender.cs:
```diff
@@ -0,0 +1,64 @@
+            .WithCredentials(_mqttConfig?.UserId, _mqttConfig?.Password)
+            .Build();
+
+        var connectResult = await mqttClient.ConnectAsync(options, CancellationToken.None);
```
Why no cancellation token?

BeeHiveJava commented on NetDaemon.Extensions.MqttEntities/MqttConfiguration.cs:
```diff
@@ -0,0 +1,9 @@
+
+internal class MqttConfiguration
+{
+    public string? Host { get; set; }
```
Are these properties really nullable?

skotl replied:
Probably not - I'll update - thanks :)

FrankBakkerNl replied:
You can probably use a record here with a primary constructor and have the properties auto generated. That way you can make then non nullabe without getting warnings. Username and password might actually be optional

FrankBakkerNl commented on NetDaemon.Extensions.MqttEntities/MessageSender.cs:
```diff
@@ -0,0 +1,64 @@
+
+    public async Task SendMessageAsync(string topic, string payload)
+    {
+        using (var mqttClient = _mqttFactory.CreateMqttClient())
```
Maybe cache this client in a field so it is not recreated each time and then dispose in Dispose if this class.

FrankBakkerNl commented on NetDaemon.Extensions.MqttEntities/DependencyInjectionSetup.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        services.AddSingleton<IMqttFactory, MqttFactory>();
+        services.AddSingleton<IEntityUpdater, EntityUpdater>();
+        services.AddSingleton<IMessageSender, MessageSender>();
```
The MessageSender does not have any state so it does not have to be a singleton. I read its recimended to use transient in those cases (creating a new instance is relatively cheap)

helto4real reviewed (approved):
I am ok for this to be merged to the feature branch so @eugeneniemand and @skotl can work on it. I save my comments to the real PR. You can use these comments as guidance for your work. Thanks for the contribution!

skotl commented on NetDaemon.Extensions.MqttEntities/IMessageSender.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace NetDaemon.Extensions.MqttEntities;
+
+public interface IMessageSender
```
Yes - definitely should be internal!
